crates/service-client/src/proxy.rs

```diff
@@ -10,18 +10,87 @@
 
 use hyper::Uri;
 use restate_types::config::ProxyUri;
-use std::task::{Context, Poll};
+use rustls::pki_types::IpAddr;
+use std::{
+    collections::HashSet,
+    task::{Context, Poll},
+};
 use tower_service::Service;
 
 #[derive(Clone, Debug)]
 pub struct ProxyConnector<C> {
     proxy: Option<ProxyUri>,
+    no_proxy_ips: HashSet<IpAddr>,
+    no_proxy_domains: Vec<String>,
     connector: C,
 }
 
 impl<C> ProxyConnector<C> {
-    pub fn new(proxy: Option<ProxyUri>, connector: C) -> Self {
-        Self { proxy, connector }
+    pub fn new(proxy: Option<ProxyUri>, no_proxy: Vec<http::uri::Authority>, connector: C) -> Self {
+        let mut no_proxy_ips = HashSet::new();
+        let mut no_proxy_domains = Vec::new();
+
+        for no_proxy_authority in no_proxy {
+            match IpAddr::try_from(no_proxy_authority.as_str()) {
+                Ok(ip) => {
+                    no_proxy_ips.insert(ip);
+                }
+                Err(_) => no_proxy_domains.push(no_proxy_authority.host().to_owned()),
+            }
+        }
+
+        Self {
+            proxy,
+            no_proxy_ips,
+            no_proxy_domains,
+            connector,
+        }
+    }
+
+    fn no_proxy(&self, host: &str) -> bool {
+        // According to RFC3986, raw IPv6 hosts will be wrapped in []. So we need to strip those off
+        // the end in order to parse correctly
+        let authority = if host.starts_with('[') && host.ends_with('[') {
```

Comment:
uuuuu maybe should be this?

```suggestion
        let authority = if host.starts_with('[') && host.ends_with(']') {
```

Reply:
ah, probably!